### PR TITLE
[RFC] Early prototype of a clock adapter.

### PR DESCRIPTION
**Pull request by Yoric**
First RFC of a clock adapter. At this stage, the adapter is not connected to the REST API. Indeed, the main idea is to test traits `Adapt` and `Adapter`.

I hope to then be able to implement trait `Adapt` into something that will expose the clock as a REST (and later WebSocket) API.

### CODE REVIEW

**Comment by Yoric** [comment]
Advancing towards the integration of Taxonomy in the FoxBox.

**Comment by Yoric** [comment]
Reorganized the patch, added plenty of doc and a few tests.

File: src/adapt.rs [hunk 1]
```diff
@@ -0,0 +1,481 @@
+//! An API for plugging in adapters.
```
Comment (fabricedesre) [inline src/adapt.rs:1]: nit: rename this file to adapters.rs ?

File: src/adapt.rs [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+#![allow(dead_code)] // Implementation in progress, code isn't called yet.
+
+use transact::{InsertInMap, InsertInSet};
```
Comment (fabricedesre) [inline src/adapt.rs]: We use the transact:: { InsertInMap, InsertInSet } style everywhere else in foxbox. I don't know if the "no space" version is the commonly used style in the rust community, but I find it less readable. In any case, we need to choose one style and stick to it.

File: src/adapt.rs [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+    fn vendor(&self) -> &str;
+    fn version(&self) -> &[u32;4];
+    // ... more metadata
+
+    /// Request a value from a channel. The FoxBox (not the adapter)
+    /// is in charge of keeping track of the age of values.
+    fn get_values(&self, set: Vec<Id<Getter>>) -> Vec<Result<Option<Value>, Error>>;
+
+    /// Request that a value be sent to a channel.
+    fn put_values(&self, values: Vec<(Id<Setter>, Value)>) -> Result<(), Error>;
```
Comment (fabricedesre) [inline src/adapt.rs]: s/put_values/set_values to be consistent with Setter ?

File: src/adapt.rs [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    fn register_watch(&self, id: Id<Getter>, threshold: Option<Value>, cb: Box<Fn(Value) + Send>) -> Result<Box<WatchGuard>, Error>;
+}
+
+struct AdapterData {
+    /// The implementation of the adapter.
+    adapter: Box<Adapter>,
+
+    /// The services for this adapter.
+    services: HashSet<Id<ServiceId>>,
+}
```
Comment (fabricedesre) [inline src/adapt.rs:77]: nit: here and in a few other places: add blank line

File: src/adapt.rs [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+                adapters: HashMap::new(),
+                services: HashMap::new(),
+                getters: HashMap::new(),
+                setters: HashMap::new(),
+            }))
+        }
+    }
+
+    /// Auxiliary function to remove an adapter, once the mutex has been acquired.
+    /// Clients should rather use Adapt::remove_adapter.
```
Comment (fabricedesre) [inline src/adapt.rs]: nit: here and in other comments: s/Adapt::_/AdapterControl::_

File: src/adapt.rs [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+    fn aux_remove_adapter(id: Id<AdapterId>, state: &mut AdapterControlState) -> Result<(), Error> {
+        let adapters = &mut state.adapters;
+        let services = &mut state.services;
+        let getters = &mut state.getters;
+        let setters = &mut state.setters;
```
Comment (fabricedesre) [inline src/adapt.rs]: It's a bit annoying that we have a flat structure with no direct composition of adapter->1..N service->(1..N getter, 1..N setter). Would that not simplify a lot these functions?

File: src/adapters/clock/mod.rs [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+use foxbox_taxonomy::services::*;
+use foxbox_taxonomy::util::Id;
+
+use std::collections::{HashMap, HashSet};
+
+use chrono;
+use chrono::*;
+use timer;
+
+static ADAPTER_ID: &'static str = "clock@link.mozilla.org";
```
Comment (fabricedesre) [inline src/adapters/clock/mod.rs]: xpcom, the come back of the contract id! :)

File: src/adapters/clock/mod.rs [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+static ADAPTER_NAME: &'static str = "Clock adapter (built-in)";
+static ADAPTER_VENDOR: &'static str = "team@link.mozilla.org";
+static ADAPTER_VERSION: [u32;4] = [0, 0, 0, 0];
+
+const SERVICE_CLOCK_ID: &'static str = "service:clock@link.mozilla.org";
+
+const GETTER_TIMESTAMP_ID: &'static str = "getter:timestamp.clock@link.mozilla.org";
+const GETTER_TIME_OF_DAY_ID: &'static str = "getter:timeofday.clock@link.mozilla.org";
```
Comment (fabricedesre) [inline src/adapters/clock/mod.rs]: Do these have to be global constants? or only unique for their service?

File: src/main.rs [hunk 9]
```diff
@@ -53,7 +56,8 @@ extern crate regex;
 // Need to be declared first so to let the macros be visible from other modules.
 #[macro_use]
 mod utils;
-
+mod transact;
```
Comment (fabricedesre) [inline src/main.rs:61]: please move it in alphabetical order

**Comment by fabricedesre** [comment]
The `Clock` example is not ideal since it merges the adapter with a singleton service. I would have preferred to see something closer to the 1 adpater - N services model that we'll have to implement for the Hue and the IP camera cases.
But I don't want to wait for ever to land that either, so it's fine to for me to land and to iterate.

File: src/adapt.rs (hunk 1 above)
Comment (Yoric) [inline src/adapt.rs:1]: Well, there is already a module `adapters`. I could move this in `adapters/mod.rs`, but that would make that file too large for my tastes. I could also move this to, say, `adapters/api.rs`, if you prefer.

File: src/adapt.rs (hunk 6 above)
Comment (Yoric) [inline src/adapt.rs]: Well, `services` already contains getters and setters. Here, these maps are designed to go the other way round, from getter/setter to service, from service to adapter. I'll see if I can improve that.

**Comment by Yoric** [comment]
Applied feedback and updated to the latest version of the Taxonomy API. Once tests are green, I'll merge.

File: src/adapt.rs (hunk 1 above)
Comment (fabricedesre) [inline src/adapt.rs:1]: I prefer adapters/api.rs

**Pull request merged**